gtsam/discrete/TableFactor.cpp

```diff
@@ -0,0 +1,415 @@
+    cardinalities_.insert(dkey);
+    denom /= dkey.second;
+    denominator_.insert(std::pair<Key, size_t>(dkey.first, denom));
+  }
+}
+
+/* ************************************************************************ */
+Eigen::SparseVector<double> TableFactor::Convert(
+    const std::vector<double>& table) {
+  Eigen::SparseVector<double> sparse_table(table.size());
+  // counting number of nonzero elements in table and reserving the space
+  const size_t nnz = std::count_if(table.begin(), table.end(),
+                                   [](size_t i) { return i != 0; });
+  sparse_table.reserve(nnz);
+  for (size_t i = 0; i < table.size(); i++) {
+    if (table[i] != 0) sparse_table.insert(i) = table[i];
+  }
+  return sparse_table;
+}
+
+/* ************************************************************************ */
+Eigen::SparseVector<double> TableFactor::Convert(const std::string& table) {
+  // Convert string to doubles
+  std::vector<double> ys;
+  std::istringstream iss(table);
+  std::copy(std::istream_iterator<double>(iss), std::istream_iterator<double>(),
+            std::back_inserter(ys));
+  return Convert(ys);
+}
+
+/* ************************************************************************ */
+bool TableFactor::equals(const DiscreteFactor& other, double tol) const {
+  if (!dynamic_cast<const TableFactor*>(&other)) {
+    return false;
+  } else {
+    const auto& f(static_cast<const TableFactor&>(other));
+    return (sparse_table_ - f.sparse_table_).sum() == 0;
+    // return sparse_table_ == f.sparse_table_;
+  }
+}
+
+/* ************************************************************************ */
+void TableFactor::print(const string& s, const KeyFormatter& formatter) const {
+  cout << s;
+  cout << " f[";
+  for (auto&& key : keys())
+    cout << boost::format(" (%1%,%2%),") % formatter(key) % cardinality(key);
+  cout << " ]" << endl;
+}
+/* ************************************************************************ */
+double TableFactor::operator()(const DiscreteValues& values) const {
+  // Find index of value in O(N) where N is number of keys
+  size_t idx = findIndex(values);
+  return sparse_table_.coeff(idx);
+}
+
+/* ************************************************************************ */
+DecisionTreeFactor TableFactor::operator*(const DecisionTreeFactor& f) const {
+  return toDecisionTreeFactor() * f;
+}
+
+/* ************************************************************************ */
+TableFactor TableFactor::operator*(const TableFactor& f) const {
+  DiscreteKeys keys = unionKeys(f);
+  size_t cardinality = 1;
+  // Create new TableFactor
+  for (DiscreteKey key : keys) cardinality *= key.second;
+  Eigen::SparseVector<double> new_sparse_table_(cardinality);
+  new_sparse_table_.reserve(sparse_table_.nonZeros() *
+                            (cardinality / sparse_table_.size()));
+  TableFactor multiplied_f(keys, new_sparse_table_);
+  for (Eigen::SparseVector<double>::InnerIterator it(f.sparse_table_); it;
```

Comment:
maybe say `using Eigen::SparseVector<double>::InnerIterator` at top of this file so you can just use `InnerIterator`....

Reply:
"Error: a class qualified name is not allowed". Instead, I `typedef` it: "typedef Eigen::SparseVector<double>::InnerIterator sparse_it;"